gap.go

```diff
@@ -311,94 +365,152 @@ func (buf *rawAdvertisementPayload) addFromOptions(options AdvertisementOptions)
 	// Check whether the field can fit this manufacturer data.
-	fieldLength := len(value) + 4
+	fieldLength := 1 + 1 + 2 + len(data) // 1 byte length, 1 byte ad type, 2 bytes uuid, actual manufacturer data
 	if int(buf.len)+fieldLength > len(buf.data) {
-		return false
+		return errAdvertisementPacketTooBig
 	}
 
 	// Add the data.
-	buf.data[buf.len+0] = uint8(fieldLength - 1)
-	buf.data[buf.len+1] = 0xff
-	buf.data[buf.len+2] = uint8(key)
-	buf.data[buf.len+3] = uint8(key >> 8)
-	copy(buf.data[buf.len+4:], value)
+	buf.data[buf.len+0] = byte(fieldLength - 1)
+	buf.data[buf.len+1] = 0xFF
+	buf.data[buf.len+2] = byte(uuid.Get16Bit())
+	buf.data[buf.len+3] = byte(uuid.Get16Bit() >> 8)
+	copy(buf.data[buf.len+4:], data)
 	buf.len += uint8(fieldLength)
 
-	return true
+	return nil
+}
+
+// addServiceData adds service data ([]byte) entries to the advertisement payload.
+func (buf *rawAdvertisementPayload) addServiceData(uuid UUID, data []byte) (err error) {
+	switch {
+	case uuid.Is16Bit(): {
+		// check if it fits
+		fieldLength := 1 + 1 + 2 + len(data) // 1 byte length, 1 byte ad type, 2 bytes uuid, actual service data
+		if int(buf.len)+fieldLength > len(buf.data) {
+			return errAdvertisementPacketTooBig 
+		}
+		// Add the data.
+		buf.data[buf.len+0] = byte(fieldLength - 1)
+		buf.data[buf.len+1] = 0x16
+		buf.data[buf.len+2] = byte(uuid.Get16Bit())
+		buf.data[buf.len+3] = byte(uuid.Get16Bit()>>8)
+		copy(buf.data[buf.len+4:], data)
+		buf.len += uint8(fieldLength)
+
+	}
+	case uuid.Is32Bit(): {
+		// check if it fits
+		fieldLength := 1 + 1 + 4 + len(data) // 1 byte length, 1 byte ad type, 4 bytes uuid, actual service data
+		if int(buf.len)+fieldLength > len(buf.data) {
+			return errAdvertisementPacketTooBig 
+		}
+		// Add the data.
+		buf.data[buf.len+0] = byte(fieldLength - 1)
+		buf.data[buf.len+1] = 0x20
+		buf.data[buf.len+2] = byte(uuid.Get32Bit())
+		buf.data[buf.len+3] = byte(uuid.Get32Bit()>>8)
+		buf.data[buf.len+4] = byte(uuid.Get32Bit()>>16)
+		buf.data[buf.len+5] = byte(uuid.Get32Bit()>>24)
+		copy(buf.data[buf.len+6:], data)
+		buf.len += uint8(fieldLength)
+	}
+	default: { // must be 128-bit uuid
+		// check if it fits
+		fieldLength := 1 + 1 + 16 + len(data) // 1 byte length, 1 byte ad type, 16 bytes uuid, actual service data
+		if int(buf.len)+fieldLength > len(buf.data) {
+			return errAdvertisementPacketTooBig 
+		}
+		// Add the data.
+		buf.data[buf.len+0] = byte(fieldLength - 1)
+		buf.data[buf.len+1] = 0x21
+		uuid_bytes := uuid.Bytes()
+		copy(buf.data[buf.len+2:], uuid_bytes[:])
+		copy(buf.data[buf.len+2+16:], data)
+		buf.len += uint8(fieldLength)
+	}
+	}
+	return nil
 }
 
 // addFlags adds a flags field to the advertisement buffer. It returns true on
 // success (the flags can be added) and false on failure.
-func (buf *rawAdvertisementPayload) addFlags(flags byte) (ok bool) {
+func (buf *rawAdvertisementPayload) addFlags(flags byte) (err error) {
```

Comment:
I think returning a boolean is fine here? It's an internal function, and the only way it could fail is when there's not enough space in the buffer.

(Same goes for other similar methods).